tsbridge/metric_config.go

```diff
@@ -0,0 +1,157 @@
+type InfluxDBMetricConfig struct {
+	SourceMetricConfig    `yaml:"_,inline"`
+	influxdb.MetricConfig `yaml:"_,inline"`
+}
+
+// Metrics returns a list of metrics defined in the configuration file.
+func (c *MetricConfig) Metrics() []*Metric {
+	return c.metrics
+}
+
+// NewMetricConfig reads and validates a configuration file, returning the MetricConfig struct.
+func NewMetricConfig(ctx context.Context, config *Config, storage storage.Manager) (*MetricConfig, error) {
+	data, err := ioutil.ReadFile(config.Options.Filename)
+	if err != nil {
+		return nil, err
+	}
+	c := &MetricConfig{}
+	if err := yaml.UnmarshalStrict(data, c); err != nil {
+		return nil, err
+	}
+
+	if err := validator.Validate(c); err != nil {
+		return nil, fmt.Errorf("configuration file validation error: %s", err)
+	}
+
+	destinations := make(map[string]string)
+	for _, d := range c.StackdriverDestinations {
+		if _, ok := destinations[d.Name]; ok {
+			return nil, fmt.Errorf("configuration file contains several destinations named '%s'", d.Name)
+		}
+		if d.ProjectID == "" {
+			d.ProjectID = projectID()
+		}
+		if d.ProjectID == "" {
+			return nil, fmt.Errorf("please provide project_id for destination '%s'", d.Name)
+		}
+		destinations[d.Name] = d.ProjectID
+	}
+
+	// Map used to ensure that metric names are unique.
+	metrics := make(map[string]bool)
+	// Function to create a new source metric, and to add it to the current configuration.
+	addSourceMetric := func(name, dest string, sourceMetric SourceMetric) error {
+		project, ok := destinations[dest]
+		if !ok {
+			return fmt.Errorf("destination '%s' not found", dest)
+		}
+		metric, err := NewMetric(ctx, name, sourceMetric, project, storage)
+		if err != nil {
+			return fmt.Errorf("cannot create metric '%s': %v", name, err)
+		}
+
+		c.metrics = append(c.metrics, metric)
+		if metrics[name] {
+			return fmt.Errorf("duplicate metric name '%s'", name)
+		}
+		metrics[name] = true
+		return nil
+	}
+
+	for _, m := range c.DatadogMetrics {
+		metric, err := datadog.NewSourceMetric(m.Name, &m.MetricConfig, config.Options.MinPointAge, config.Options.CounterResetInterval)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create Datadog source metric '%s': %v", m.Name, err)
+		}
+
+		if err = addSourceMetric(m.Name, m.Destination, metric); err != nil {
+			return nil, err
+		}
+	}
+
+	for _, m := range c.InfluxDBMetrics {
+		metric, err := influxdb.NewSourceMetric(m.Name, &m.MetricConfig, config.Options.MinPointAge, config.Options.CounterResetInterval)
+		if err != nil {
+			return nil, fmt.Errorf("cannot create InfluxDB source metric '%s': %v", m.Name, err)
+		}
+
+		if err = addSourceMetric(m.Name, m.Destination, metric); err != nil {
+			return nil, err
+		}
+	}
+
+	log.WithContext(ctx).Debugf("Read %d metrics and %d destinations from the config file", len(metrics), len(destinations))
+	return c, nil
+}
+
+// projectID returns the name of the GCP project that code is running in.
+func projectID() string {
```

Comment:
Do you need this now when you have `env.AppEngineProject()`?

Reply:
Agreed, fixed. Just had to change the `env` module a little bit to make sure we don't run into confusion issues in the future should AppEngine change env.